smdebug/core/hook.py

```diff
@@ -652,8 +661,7 @@ def _write_scalars(self):
                         scalar_name, scalar_val, self.step, timestamp=timestamp
                     )
             if write_event:
-                if self.writer is None:
-                    self._initialize_writers()
+                self._initialize_writers(only_initialize_if_missing=True)
```

Comment:
how is this different that what was there earlier.
it will be no op if self.writer is not None which is same as earlier behavior.

Reply:
`self.writer` is always `None` for mirrored strategy.

So, in the earlier case, `self._initialize_writers()` was always creating new writers and deleting the existing writers with buffered data.

`self._initialize_writers(only_initialize_if_missing=True)` calls https://github.com/awslabs/sagemaker-debugger/blob/e99b7054f4be814c751890344969a825f257f9cb/smdebug/tensorflow/base_hook.py#L305

which handles the mirrored strategy case

Reply:
say -  if self.writer is None for mirrored strategy, 
This line         if only_initialize_if_missing and self.writer: ==> will evaluate to false. and then self.writer would be inited in L449, 
The same thing will happen if this code change is not there. Am i missing anything here.

Reply:
Yes, that is correct.
But before my code change, we had two 
`_initialize_writers` fns. One that accepted only_initialize_if_missing and one that didn't.
I have simply just unified the two of them.

Reply:
Here is the snippet of my change to the `_initialize_writers` fn:

```
    def _initialize_writers(self, only_initialize_if_missing=False) -> None:
        if only_initialize_if_missing and self.writer:
            return
```

Reply:
Link to the impl in base_hook.

https://github.com/awslabs/sagemaker-debugger/blob/e6359444d28ae7de260074ef4d5962ac83fa649a/smdebug/tensorflow/base_hook.py#L305

Reply:
can you write a comment on core hook initialize writers that this method can be overridden. For tensorflow, the method is overridden in baseHook class.